package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -0,0 +1,688 @@
+    return a[XX]*a[XX]+a[YY]*a[YY]+a[ZZ]*a[ZZ]
+
+cdef void rvec_clear(rvec a) nogil:
+    a[XX]=0.0
+    a[YY]=0.0
+    a[ZZ]=0.0
+
+###############################
+# Utility class to handle PBC #
+###############################
+cdef struct cPBCBox_t:
+    matrix     box
+    rvec       fbox_diag
+    rvec       hbox_diag
+    rvec       mhbox_diag
+    real       max_cutoff2
+
+
+# Class to handle PBC calculations
+cdef class PBCBox(object):
+    cdef cPBCBox_t c_pbcbox
+    cdef rvec center
+    cdef rvec bbox_center
+    cdef bint is_triclinic
+
+    def __init__(self, real[:,::1] box):
+        self.update(box)
+
+    cdef void fast_update(self, real[:,::1] box) nogil:
+        cdef ns_int i, j, k, d, jc, kc, shift
+        cdef real d2old, d2new, d2new_c
+        cdef rvec trial, pos
+        cdef ns_int ii, jj ,kk
+        cdef ns_int *order = [0, -1, 1, -2, 2]
+        cdef bint use
+        cdef real min_hv2, min_ss, tmp
+
+        rvec_clear(self.center)
+        # Update matrix
+        self.is_triclinic = False
+        for i in range(DIM):
+            for j in range(DIM):
+                self.c_pbcbox.box[i][j] = box[i, j]
+                self.center[j] += 0.5 * box[i, j]
+
+                if i != j:
+                    if box[i, j] > EPSILON:
+                        self.is_triclinic = True
+            self.bbox_center[i] = 0.5 *  box[i, i]
+
+        # Update diagonals
+        for i in range(DIM):
+            self.c_pbcbox.fbox_diag[i]  =  box[i, i]
+            self.c_pbcbox.hbox_diag[i]  =  self.c_pbcbox.fbox_diag[i] * 0.5
+            self.c_pbcbox.mhbox_diag[i] = - self.c_pbcbox.hbox_diag[i]
+
+        # Update maximum cutoff
+
+        # Physical limitation of the cut-off
+        # by half the length of the shortest box vector.
+        min_hv2 = min(0.25 * rvec_norm2(&box[XX, XX]), 0.25 * rvec_norm2(&box[YY, XX]))
+        min_hv2 = min(min_hv2, 0.25 * rvec_norm2(&box[ZZ, XX]))
+
+        # Limitation to the smallest diagonal element due to optimizations:
+        # checking only linear combinations of single box-vectors (2 in x)
+        # in the grid search and pbc_dx is a lot faster
+        # than checking all possible combinations.
+        tmp = box[YY, YY]
+        if box[ZZ, YY] < 0:
+            tmp -= box[ZZ, YY]
+        else:
+            tmp += box[ZZ, YY]
+
+        min_ss = min(box[XX, XX], min(tmp, box[ZZ, ZZ]))
+        self.c_pbcbox.max_cutoff2 = min(min_hv2, min_ss * min_ss)
+
+    def update(self, real[:,::1] box):
+        if box.shape[0] != DIM or box.shape[1] != DIM:
+            raise ValueError("Box must be a {} x {} matrix. Got: {} x {})".format(
+                DIM, DIM, box.shape[0], box.shape[1]))
+        if (box[XX, XX] == 0) or (box[YY, YY] == 0) or (box[ZZ, ZZ] == 0):
+            raise ValueError("Box does not correspond to PBC=xyz")
+        self.fast_update(box)
+
+
+    cdef void fast_pbc_dx(self, rvec ref, rvec other, rvec dx) nogil:

```

Comment:
@ayushsuhane this just does orthogonal PBC because we back into a rectangular box.  Can this be replaced with the `minimum_image_ortho` function in calc_distances.h?

Reply:
Yes, from a first look, i think it can. But the advantage of not searching all the 27 images will be lost. I will try to do a benchmark with it first and get back here.

Reply:
Can you elaborate a little bit. I think using directly minimum_image_ortho will not give correct result. I understand that coordinates are packed into an orthogonal brick with its origin at the centre of original triclinic box. But inorder to evaluate the distances (``dx``), shifts corresponding to the triclinic box vectors are necessary (therefore defining inverse box vectors as the ``1/box[0][0]`` will not give correct result). 

![triclininc](https://user-images.githubusercontent.com/34154224/43371276-3a8cd1ec-9343-11e8-9b88-04344eb56c69.png)

What I understand is if we use the minimum_image_ortho the dx corresponding to the AD will be returned but the correct distance would be AC. Probably the better way would be to get the inverse of the box, do a dot product with the dx, and then use it with fractional coordinates as minimum_image_ortho? Am I getting it right. 

Please correct me if this is wrong. @richardjgowers @jbarnoud @kain88-de

Reply:
Ok yeah I was wrong here